src/main/java/com/minecolonies/coremod/entity/ai/citizen/deliveryman/EntityAIWorkDeliveryman.java

```diff
@@ -13,12 +30,371 @@
 public class EntityAIWorkDeliveryman extends AbstractEntityAIInteract<JobDeliveryman>
 {
     /**
+     * Min distance the worker should have to the warehouse to make any decisions.
+     */
+    private static final int MIN_DISTANCE_TO_WAREHOUSE = 5;
+
+    /**
+     * Warehouse the deliveryman is assigned to.
+     */
+    private BuildingWareHouse wareHouse = null;
+
+    /**
+     * List of itemStacks the deliveryman will deliver to the requesting building.
+     */
+    private List<ItemStack> itemsToDeliver = new ArrayList<>();
+
+    /**
+     * Next target the deliveryman should gather stuff at.
+     */
+    private BlockPos gatherTarget = null;
+
+    /**
      * Initialize the deliveryman and add all his tasks.
      *
      * @param deliveryman the job he has.
      */
     public EntityAIWorkDeliveryman(@NotNull final JobDeliveryman deliveryman)
     {
         super(deliveryman);
+        super.registerTargets(
+                /**
+                 * Check if tasks should be executed.
+                 */
+                new AITarget(this::checkIfExecute, IDLE),
+                new AITarget(IDLE, () -> START_WORKING),
+                new AITarget(START_WORKING, this::checkWareHouse),
+                new AITarget(PREPARE_DELIVERY, this::prepareDelivery),
+                new AITarget(GATHER_IN_WAREHOUSE, this::gatherItemsFromWareHouse),
+                new AITarget(DELIVERY, this::deliver),
+                new AITarget(GATHERING, this::gather),
+                new AITarget(DUMPING, this::dump)
+
+        );
+        worker.setSkillModifier(2 * worker.getCitizenData().getEndurance() + worker.getCitizenData().getCharisma());
+        worker.setCanPickUpLoot(true);
+    }
+
+    /**
+     * Gather items from a random hut which the hut doesn't need.
+     * @return the next state to go to.
+     */
+    public AIState gather()
+    {
+        if(gatherTarget == null)
+        {
+            gatherTarget = getRandomBuilding();
+        }
+
+        if(gatherTarget == null)
+        {
+            return GATHERING;
+        }
+
+        if(!worker.isWorkerAtSiteWithMove(gatherTarget, MIN_DISTANCE_TO_WAREHOUSE))
+        {
+            return GATHERING;
+        }
+
+        final Colony colony = getOwnBuilding().getColony();
+        if(colony != null)
+        {
+            final AbstractBuilding building = colony.getBuilding(gatherTarget);
+            gatherFromBuilding(building);
+            return DUMPING;
+        }
+        return START_WORKING;
+    }
+
+    /**
+     * Gather not needed Items from building.
+     * @param building building to gather it from.
+     */
+    private void gatherFromBuilding(@NotNull final AbstractBuilding building)
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "gatherFromBuilding" is 12 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)